### PR TITLE
Server: Hit Ctrl+C twice to exit

### PR DESCRIPTION
**Pull request by ngxson**
While it's nice to have the server [properly shutdown on SIGINT](https://github.com/ggerganov/llama.cpp/pull/5244), it sometimes blocks the development and force the developer to use `killall server`

This PR propose a quite sketchy solution: If the SIGINT is received the second time, we just force exit the process.

### CODE REVIEW

File: examples/server/server.cpp [hunk 1]
```diff
@@ -2768,7 +2768,14 @@ static void append_to_generated_text_from_generated_token_probs(llama_server_con
 }
 
 std::function<void(int)> shutdown_handler;
-inline void signal_handler(int signal) { shutdown_handler(signal); }
+static bool is_terminating = false;
+inline void signal_handler(int signal) {
+    if (is_terminating) {
+        exit(0); // force exit the process, in case it hangs
+    }
+    is_terminating = true;
```
Comment (cebtenzzre) [inline examples/server/server.cpp:2776]: It would be better to do this with std::atomic_flag test_and_set, since any thread can receive the signal.

Comment (ngxson) [reply examples/server/server.cpp]: Thanks for suggestion, that was done in https://github.com/ggerganov/llama.cpp/pull/5734/commits/273d985271580bce0684889d007bbac7b2484e66

File: examples/server/server.cpp [hunk 2]
```diff
@@ -2768,7 +2768,13 @@ static void append_to_generated_text_from_generated_token_probs(llama_server_con
 }
 
 std::function<void(int)> shutdown_handler;
-inline void signal_handler(int signal) { shutdown_handler(signal); }
+std::atomic_flag is_terminating = ATOMIC_FLAG_INIT;
+inline void signal_handler(int signal) {
+    if (is_terminating.test_and_set()) {
+        exit(0); // force exit the process, in case it hangs
```
Comment (cebtenzzre) [inline examples/server/server.cpp:2774]: Since this is an unclean exit, wouldn't it be better to `exit(1)`?

Comment (ngxson) [reply examples/server/server.cpp:2774]: Yeah that's reasonable, I'll do that if we all agree that it's a good idea to add this such feature to the code base.

I actually mentioned in the description that this is a sketchy solution, so I'm not even sure if it's a good idea or not. Just a small stuff to have a better developer experience.

**Review by phymbert — APPROVED** [review]

**Review by ggerganov — APPROVED** [review]

File: examples/server/server.cpp (hunk 2 above)
Comment (ngxson) [reply examples/server/server.cpp]: I added the message and exit(1), also I put a comment to say that this is a "hotfix" that can be removed when the server become more stable

File: examples/server/server.cpp [hunk 3]
```diff
@@ -2772,7 +2772,16 @@ static void append_to_generated_text_from_generated_token_probs(llama_server_con
 }
 
 std::function<void(int)> shutdown_handler;
-inline void signal_handler(int signal) { shutdown_handler(signal); }
+std::atomic_flag is_terminating = ATOMIC_FLAG_INIT;
+inline void signal_handler(int signal) {
+    if (is_terminating.test_and_set()) {
+        // in case it hangs, we can force terminate the server by hitting Ctrl+C twice
+        // this is for better developer experience, we can remove when the server is stable enough
+        fprintf(stderr, "SIGINT is received second time, exit immediately...\n");
```
Comment (cebtenzzre) [inline examples/server/server.cpp:2780]: This is somewhat broken English, let me suggest the specific wording that I used in my example:
```suggestion
        fprintf(stderr, "Received second interrupt, terminating immediately.\n");
```
(optionally replace "interrupt" with "SIGINT" and/or "terminating" with "exiting")

**Pull request merged**